src/main/scala/mesosphere/marathon/storage/migration/MigrationTo160.scala

```diff
@@ -0,0 +1,127 @@
+package mesosphere.marathon
+package storage.migration
+
+import java.time.OffsetDateTime
+
+import akka.Done
+import akka.http.scaladsl.unmarshalling.Unmarshaller
+import akka.stream.Materializer
+import akka.stream.scaladsl.Sink
+import com.typesafe.scalalogging.StrictLogging
+import mesosphere.marathon.core.instance.{ Instance, Reservation }
+import mesosphere.marathon.core.instance.Instance.Id
+import mesosphere.marathon.core.storage.store.impl.cache.{ LazyCachingPersistenceStore, LazyVersionCachingPersistentStore, LoadTimeCachingPersistenceStore }
+import mesosphere.marathon.core.storage.store.{ IdResolver, PersistenceStore }
+import mesosphere.marathon.core.storage.store.impl.zk.{ ZkId, ZkPersistenceStore, ZkSerialized }
+import mesosphere.marathon.storage.repository.InstanceRepository
+import play.api.libs.json.{ JsValue, Json }
+
+import scala.concurrent.{ ExecutionContext, Future }
+
+@SuppressWarnings(Array("ClassNames"))
+class MigrationTo160(instanceRepository: InstanceRepository, persistenceStore: PersistenceStore[_, _, _])(implicit ctx: ExecutionContext, mat: Materializer) extends StrictLogging {
+
+  def migrate(): Future[Done] = {
+    MigrationTo160.migrateReservations(instanceRepository, persistenceStore)
+  }
+}
+
+object MigrationTo160 extends StrictLogging {
+  /**
+    * This function traverses all instances in ZK, and moves reservation objects from tasks to the instance level.
+    */
+  def migrateReservations(instanceRepository: InstanceRepository, persistenceStore: PersistenceStore[_, _, _])(implicit ctx: ExecutionContext, mat: Materializer): Future[Done] = {
+
+    logger.info("Starting reservations migration to 1.6.0")
+
+    /**
+      * We're trying to find if we have a ZooKeeper store because it provides objects as byte arrays and this
+      * makes serialization into json easier.
+      */
+    val maybeStore: Option[ZkPersistenceStore] = {
+
+      def findZkStore(ps: PersistenceStore[_, _, _]): Option[ZkPersistenceStore] = {
+        ps match {
+          case zk: ZkPersistenceStore =>
+            Some(zk)
+          case lcps: LazyCachingPersistenceStore[_, _, _] =>
+            findZkStore(lcps.store)
+          case lvcps: LazyVersionCachingPersistentStore[_, _, _] =>
+            findZkStore(lvcps.store)
+          case ltcps: LoadTimeCachingPersistenceStore[_, _, _] =>
+            findZkStore(ltcps.store)
+          case other =>
+            None
+        }
+      }
+
+      findZkStore(persistenceStore)
+    }
+
+    implicit val instanceResolver: IdResolver[Instance.Id, JsValue, String, ZkId] =
```

Comment:
Why is this required?

Reply:
This is an implicit parameter for zk store

Reply:
Don't we have this defined already?

Reply:
This one is for operating with `JsValue`s.